spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -207,7 +207,14 @@ private List<FileScanTask> planFiles(StreamingOffset startOffset, StreamingOffse
         currentOffset = new StreamingOffset(snapshotAfter.snapshotId(), 0L, false);
       }
 
-      if (!shouldProcess(table.snapshot(currentOffset.snapshotId()))) {
+      Snapshot snapshot = table.snapshot(currentOffset.snapshotId());
+
+      if (snapshot == null) {
+        throw new IllegalStateException(
+            String.format("Failed to find expected snapshot %d", snapshot.snapshotId()));
+      }
+
```

Comment:
Will write a test for this case, if we determine if it's the right way to handle this case.

Reply:
I'm curious whether we could just add this to `shouldProcess`:

```
   private boolean shouldProcess(Snapshot snapshot) {
+    if (null == snapshot) {
+      return false;
+    }
     String op = snapshot.operation();
     switch (op) {
       case DataOperations.APPEND:
```

Reply:
agreed, if a snapshot is not found in the table, it's bad. It indicates the reader's checkpoint is too old. There is no nice way to recover from this, except clearing your checkpoints and restarting.

this happen to me because I was clearing my snapshots every hour. I realized then that this made no sense and that I should keep a week of snapshot. Readers have up to a week to fix any streaming job which went down.

So I'm all good with throwing an exception that points the user to the fact that the checkpointed snapshot ID is no longer available in the table.